devito/propagator.py

```diff
@@ -197,10 +197,9 @@ def sympy_to_cgen(self, stencils):
         if len(self.factorized) > 0:
             for name, term in zip(self.factorized.keys(), self.factorized):
                 expr = self.factorized[name]
-                # TODO: add support for double precision
-                self.add_local_var(name, "float")
-                # TODO: undo precision enforcing
-                factors.append(cgen.Assign(name, str(ccode(expr.xreplace(self._var_map))).replace("pow", "powf").replace("fabs", "fabsf")))
+                self.add_local_var(name, self.dtype)
+                if self.dtype is np.float32:
+                    factors.append(cgen.Assign(name, str(ccode(expr.xreplace(self._var_map))).replace("pow", "powf").replace("fabs", "fabsf")))
```

Comment:
I don't think its a good idea to be doing replaces on the generated C code. What's worse, this is already in master. We should be hooking in to sympy codegenerator to manipulate this at the time of code generation instead. This would remove the need for Propagator to be aware of the dtype. The less each class is aware of the others, the more general they are.